utils/pattern/src/single.rs

```diff
@@ -0,0 +1,256 @@
+
+    fn validate_store(store: &Self::Store) -> Result<(), Error> {
+        let placeholder_offset_char = store.chars().next().ok_or(Error::InvalidPattern)?;
+        let initial_offset = char::len_utf8(placeholder_offset_char);
+        let placeholder_offset = placeholder_offset_char as usize;
+        if placeholder_offset > store.len() - initial_offset + 1 {
+            return Err(Error::InvalidPattern);
+        }
+        Ok(())
+    }
+
+    fn iter_items(store: &Self::Store) -> Self::Iter<'_> {
+        let placeholder_offset_char = match store.chars().next() {
+            Some(i) => i,
+            None => {
+                debug_assert!(false);
+                '\0'
+            }
+        };
+        let initial_offset = char::len_utf8(placeholder_offset_char);
+        SinglePlaceholderPatternIterator {
+            store,
+            placeholder_offset: placeholder_offset_char as usize + initial_offset - 1,
+            current_offset: initial_offset,
+        }
+    }
+
+    #[cfg(feature = "alloc")]
+    fn try_from_items<
+        'a,
+        I: Iterator<Item = Result<PatternItemCow<'a, Self::PlaceholderKey>, Error>>,
+    >(
+        items: I,
+    ) -> Result<String, Error> {
+        let mut result = String::new();
+        let mut seen_placeholder = false;
+        for item in items {
+            match item? {
+                PatternItemCow::Literal(s) => result.push_str(&s),
+                PatternItemCow::Placeholder(_) if !seen_placeholder => {
+                    seen_placeholder = true;
+                    let placeholder_offset =
+                        u32::try_from(result.len() + 1).map_err(|_| Error::InvalidPattern)?;
+                    let placeholder_offset_char =
+                        char::try_from(placeholder_offset).map_err(|_| Error::InvalidPattern)?;
+                    result.insert(0, placeholder_offset_char);
+                }
+                PatternItemCow::Placeholder(_) => {
+                    return Err(Error::InvalidPattern);
+                }
+            }
+        }
+        if !seen_placeholder {
+            result.insert(0, '\0');
+        }
+        Ok(result)
+    }
+}
+
+#[doc(hidden)] // TODO(#4467): Should be internal
+#[derive(Debug)]
+pub struct SinglePlaceholderPatternIterator<'a> {
+    store: &'a str,
+    placeholder_offset: usize,
+    current_offset: usize,
+}
+
+impl<'a> Iterator for SinglePlaceholderPatternIterator<'a> {
+    type Item = PatternItem<'a, SinglePlaceholderKey>;
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.current_offset.cmp(&self.placeholder_offset) {
+            Ordering::Less => {
+                // Prefix
+                let result = self
+                    .store
+                    .get(self.current_offset..self.placeholder_offset)
+                    .map(PatternItem::Literal);
+                self.current_offset = self.placeholder_offset;
+                result
```

Comment:
issue: your invariant guarantees that the `get` succeeds, so for the benefit of the next person, this should be locally obvious. Use index slicing and construct an explicit `Some`

Reply:
Done